File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-3.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {

        ListNode helper = new ListNode(-1, head);
        ListNode left   = helper;
        ListNode right  = head;

        while ( n > 0 ) {
            // Traverse n nodes from linkedlist with right pointer
            right = right.next;
            n--;
        }

        while ( right != null ) {
            // Traverse linkedlist with both left and right pointer
            left  = left.next;
            right = right.next;
        }

        // Skip node of linkedlist
        left.next = left.next.next;

        return helper.next;
    }
}

// class Solution {
//     public ListNode removeNthFromEnd(ListNode head, int n) {

//         ListNode curr = head;
//         int size = 0;

//         while ( curr != null ) {
//             curr = curr.next;
//             size++;
//         }

//         if ( n == size ) return head.next;

//         // Reset curr node
//         curr = head;
//         // Traverse 
//         while ( size - n - 1 > 0 ) {
//             curr = curr.next;
//             size--;
//         }
//         // Skip node
//         curr.next = curr.next.next;

//         return head;

//     }
// }

File: Data Structures & Algorithms/reverse-a-linked-list/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        prev, curr = None, head

        while curr:
            # A new Node after curr Node
            temp = curr.next
            # Reverse the pointer of curr node to oposite direction
            curr.next = prev
            # Prev node traverses linkedlist by 1 step
            prev = curr
            # Curr node traverses linkedlist by 1 step
            curr = temp


        return prev 
        